include/revng/Pipeline/Step.h

```diff
@@ -0,0 +1,387 @@
+                             Containers &Containers,
+                             const llvm::ArrayRef<std::string> &ArgsNames) {
+  constexpr auto
+    Indexes = std::make_integer_sequence<size_t, sizeof...(Args)>();
+  revng_assert(sizeof...(Args) == ArgsNames.size());
+  invokeImpl(Ctx, Enf, F, Containers, ArgsNames, Indexes);
+}
+
+class PipeWrapperBaseImpl {
+public:
+  virtual void run(Context &Ctx, Containers &Containers) = 0;
+  virtual ContainersStatus
+  getRequirements(const ContainersStatus &Target) const = 0;
+  virtual ContainersStatus deduceResults(ContainersStatus &Target) const = 0;
+  virtual ~PipeWrapperBaseImpl() = default;
+  virtual std::unique_ptr<PipeWrapperBaseImpl> clone() const = 0;
+  virtual std::vector<std::string> getRunningContainersNames() const = 0;
+  virtual std::string getName() const = 0;
+  virtual void dump(std::ostream &OS, size_t indents) const = 0;
+};
+
+template<typename T>
+concept Printable = requires(T a) {
+  { a.print(dbg, 0) };
+};
+
+template<typename Pipe>
+class PipeWrapperImpl : public PipeWrapperBaseImpl {
+private:
+  Pipe Enf;
+  std::vector<std::string> RunningContainersNames;
+
+public:
+  PipeWrapperImpl(Pipe Enf, std::vector<std::string> RunningContainersNames) :
+    Enf(std::move(Enf)),
+    RunningContainersNames(std::move(RunningContainersNames)) {}
+
+  void run(Context &Ctx, Containers &Containers) override {
+    invokeEnforcingFunction(Ctx,
+                            Enf,
+                            &Pipe::run,
+                            Containers,
+                            RunningContainersNames);
+  }
+
+  ContainersStatus
+  getRequirements(const ContainersStatus &Target) const override {
+    const auto &Contracts = Enf.getContract();
+    auto ToReturn = Target;
+    for (const auto &Contract : llvm::reverse(Contracts))
+      ToReturn = Contract.deduceRequirements(ToReturn, RunningContainersNames);
+    return ToReturn;
+  }
+
+  ContainersStatus deduceResults(ContainersStatus &Target) const override {
+    const auto &Contracts = Enf.getContract();
+    for (const auto &Contract : Contracts)
+      Contract.deduceResults(Target, RunningContainersNames);
+    return Target;
+  }
+
+  ~PipeWrapperImpl() override = default;
+
+  std::unique_ptr<PipeWrapperBaseImpl> clone() const override {
+    return std::make_unique<PipeWrapperImpl>(*this);
+  }
+
+  std::vector<std::string> getRunningContainersNames() const override {
+    return RunningContainersNames;
+  }
+
+  std::string getName() const override { return Pipe::Name; }
+
+  void dump(std::ostream &OS, size_t Indents) const override {
+    if constexpr (Printable<Pipe>)
+      Enf.dump(OS, Indents);
+  }
+};
+
+} // namespace Detail
+
+/// A pipe must be type erased somehow to become compatible with a
+/// pipeline, a PipeWrapper takes care of this issue, it can be
+/// constructed from any pipeline type, and it will expose the
+/// contract and run method of that enforcer.
```

Comment:
```suggestion
/// A pipe must be type erased somehow to become compatible with a
/// pipeline, a PipeWrapper takes care of this issue, it can be
/// constructed from any pipe type, and it will expose the
/// contract and run method of that pipe.
```
Please review all the usages of the word *enforcer*.